fix: ensure we capture all state updates from controllers

davidmurdoch opened the pull request:
## **Description**

Controllers like to update state during during Init function calls, and we could potentially miss these changes. This doesn't cause any problems today, as we always end up with an `update` call later, which write all state.

In the [near future](https://github.com/MetaMask/metamask-extension/pull/37060) we'll only write what has changed. So we need to be sure we capture any and all changes, in case the controllers that changed in their `Init` functions aren't ever changed again -- as we'll never revisit their state to save it again.

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/38784?quickstart=1)

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry: prepare storage listener for split state storage

<!--

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**


### **Before**


### **After**

-->

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Persist state changes detected during controller bootstrap and refactor MetaMaskController to configure controllers via an idempotent method using instance-scoped version and options.
> 
> - **Background (`app/scripts/background.js`)**:
>   - Detect and persist controller state deltas immediately after instantiation by diffing `initState` vs `controller.store.getState()`; log and call `update` if changes exist.
>   - Register persistence hooks (`controller.store.on('update'|'error')`) before calling `controller.configureAllControllers`.
> - **Controller (`app/scripts/metamask-controller.js`)**:
>   - Introduce `configureAllControllers()` (idempotent) and move all controller initialization into it; call from constructor.
>   - Store version on instance (`this.version`) and use it for `web3_clientVersion`, `version` field, and first-time info; expose via `getVersion`.
>   - Use `this.opts.initState` and `this.opts.isFirstMetaMaskControllerSetup` where applicable.
>   - Initialize multiple-instance detection earlier; remove duplicate later listener setup.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 420041d5a6e52d6e7447d3615fc67d859879e4f9. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

davidmurdoch committed:
fix: ensure we capture all state updates from controllers

davidmurdoch commented on app/scripts/background.js:
```diff
@@ -736,12 +736,6 @@ async function initialize(backup) {
-    log.error('MetaMask controller.store error:', error);
-    sentry?.captureException(error);
-  });
-
```
moved inside `setupController`

davidmurdoch commented on app/scripts/background.js:
```diff
@@ -1144,8 +1138,22 @@ export function setupController(
+  });
+
+  // `autoConfigureControllers` is `false`, so we have to call this manually here.
+  controller.configureAllControllers();
```
important that this is called after we start listening for changes.

davidmurdoch commented on app/scripts/metamask-controller.js:
```diff
@@ -490,7 +490,7 @@ export default class MetamaskController extends EventEmitter {
     this.notificationManager = opts.notificationManager;
     const initState = opts.initState || {};
-    const version = process.env.METAMASK_VERSION;
+    this.version = process.env.METAMASK_VERSION;
```
Some stuff that used this `version` value was moved out of the constructor into a function. Also noticed a few other places `process.env.METAMASK_VERSION` was set, so I updated those to use the new class property I've created here.

Gudahtt commented on app/scripts/background.js:
```diff
@@ -1144,8 +1138,22 @@ export function setupController(
 
+  // this persistence hooks must happen before `controller.configureAllControllers`
+  // or some update might not be persisted!
+  controller.store.on('update', update);
```
I think I understand why those Sentry tests are failing now. This is unsafe to do right now because we'll end up persisting partway through initialization, when not all controllers are constructed yet. This is OK after we do state splitting, but until then we're capturing a "complete" state snapshot each time (or at least attempting to), but it's not really complete because we're mid-init.

If we want to introduce this before state splitting, I think we'll need a guard against this running before controller initialization has finished.

davidmurdoch replied:
AH, this is just for the _sentry_ part. Persistence is fine, right?

Is this only "unsafe" for the tests?

davidmurdoch committed:
update snapshot

davidmurdoch committed:
Merge branch 'main' into early-subscribe

davidmurdoch committed:
try this

davidmurdoch committed:
hm

Gudahtt committed:
test: Prevent double `FixtureExtensionStore` initialization

The `FixtureExtensionStore` gets initialized both in the background and
in each UI instance. Each time it reads the fixture state from disk and
writes it to `storage.local`.

The store has been updated to optionally skip initialization. It no
longer initializes by default, you need to opt-in to that step. It will
now only initialize in the background process.

davidmurdoch committed:
put things back

davidmurdoch committed:
Merge branch 'main' into early-subscribe

davidmurdoch committed:
maybe this time!

davidmurdoch committed:
Merge branch 'main' into early-subscribe

davidmurdoch committed:
try 10 seconds

davidmurdoch committed:
Merge branch 'early-subscribe' of https://github.com/MetaMask/metamask-extension into early-subscribe

davidmurdoch committed:
try this

davidmurdoch committed:
fix comment

davidmurdoch committed:
update the JSON again

davidmurdoch committed:
good catch, robot

davidmurdoch committed:
lint

davidmurdoch committed:
Merge branch 'main' into early-subscribe

davidmurdoch committed:
revert controller changes

davidmurdoch committed:
Merge branch 'early-subscribe' of https://github.com/MetaMask/metamask-extension into early-subscribe

Gudahtt reviewed (approved):
LGTM!
